File: src/Homeworks/Homework5.Tests/CalculatorTests.fs
module Tests.Calculator

open Homework5
open Xunit
open Xunit

let checkCorrectOperation firstNum secondNum operation expected =
    let actual = Calculator.tryCalculate firstNum operation secondNum
    match actual with
    | Ok o -> Assert.Equal(expected, o, 10)
    | Error e -> Assert.True(false)


let Sum: obj[][] = [|
    [|1; 2; Plus; 3|]
    [|-2; 3; Plus; 1|]
    [|10.32; 2.3; Plus; 12.62|]
    [|-23.256; 3.128; Plus; -20.128|]
|]

[<Theory>]
[<MemberData(nameof(Sum))>]
let Calculate_Sum_ReturnOk firstNum secondNum operation expected =
    checkCorrectOperation firstNum secondNum operation expected

let Subtraction: obj[][] = [|
    [|5; 2; Minus; 3|]
    [|2; -6; Minus; 8|]
    [|-3.34; 4.10; Minus; -7.44|]
    [|10.34; 3.233; Minus; 7.107|]
|]

[<Theory>]
[<MemberData(nameof(Subtraction))>]
let Calculate_Subtraction_ReturnOk firstNum secondNum operation expected =
    checkCorrectOperation firstNum secondNum operation expected
    
let Multiplication: obj[][] = [|
    [|5; 2; Multiply; 10|]
    [|10; -2; Multiply; -20|]
    [|2.3; 4.5; Multiply; 10.35|]
    [|-12.345; 10; Multiply; -123.45|]
|]

[<Theory>]
[<MemberData(nameof(Multiplication))>]
let Calculate_Multiplication_ReturnOk firstNum secondNum operation expected =
    checkCorrectOperation firstNum secondNum operation expected
    
let Division: obj[][] = [|
    [|10; 5; Divide; 2|]
    [|12; 5; Divide; 2.4|]
    [|2.4; -1.2; Divide; -2|]
    [|-13; 4; Divide; -3.25|]
|]

[<Theory>]
[<MemberData(nameof(Division))>]
let Calculate_Division_ReturnOk firstNum secondNum operation expected =
    checkCorrectOperation firstNum secondNum operation expected
    
let DivisionError: obj[][] = [|
    [|10; 0; Divide|]
    [|-12.12; 0; Divide|]
|]

[<Theory>]
[<MemberData(nameof(DivisionError))>]
let Calculate_DivisionZero_ReturnError firstNum secondNum operation =
    let actual = Calculator.tryCalculate firstNum operation secondNum
    match actual with
    | Ok o -> Assert.True(false)
    | Error e -> Assert.Equal(3, e)
    

File: src/Homeworks/Homework8.Tests/CalculatorTests.cs
using Xunit;

namespace Homework8.Tests
{
	public class CalculatorTests
	{
		private static readonly NumericalCalculator calculator = new();
		
		[Theory]
		[InlineData(1, 2, 3)]
		[InlineData(-2, 3, 1)]
		[InlineData(2, -4, -2)]
		[InlineData(-3, -4, -7)]
		public void Add_AddTwoNumbers_ReturnSum(
			int firstNum, int secondNum, int expected)
		{
			var actual = calculator.Add(firstNum, secondNum);
			Assert.Equal(expected, actual);
		}

		[Theory]
		[InlineData(5, 2, 3)]
		[InlineData(2, -6, 8)]
		[InlineData(-3, 4, -7)]
		[InlineData(-8, -2, -6)]
		public void Subtract_ReturnDifference(
			int firstNum, int secondNum, int expected)
		{
			var actual = calculator.Subtract(firstNum, secondNum);
			Assert.Equal(expected, actual);
		}

		[Theory]
		[InlineData(0, 0, 0)]
		[InlineData(0, -6, 0)]
		[InlineData(4, 0, 0)]
		[InlineData(1, 20, 20)]
		[InlineData(3, -2, -6)]
		[InlineData(-5, -3, 15)]
		public void Multiply_ReturnProduct(
			int firstNum, int secondNum, int expected)
		{
			var actual = calculator.Multiply(firstNum, secondNum);
			Assert.Equal(expected, actual);
		}

		[Theory]
		[InlineData(20, 2, 10)]
		[InlineData(10, -5, -2)]
		[InlineData(-9, -9, 1)]
		[InlineData(0, 2, 0)]
		[InlineData(0, -10, 0)]
		[InlineData(12, 0, 0)]
		[InlineData(0, 0, 0)]
		public void Divide_ReturnQuotient(
			int firstNum, int secondNum, int expected)
		{
			var actual = calculator.Divide(firstNum, secondNum);
			Assert.Equal(expected, actual);
		}
	}
}